### PR TITLE
casper-cep78-js-client 1.0.0

### LINKED ISSUE #94: Final Prep for cep78-js-sdk library to be released
- [x] Add new entrypoints support
- [x] Extend tests
- [x] Add some basic README linking to @ACStoneCL tutorial and cep78 documentation

### PR DESCRIPTION
**Pull request by hoffmannjan**
## casper-cep78-js-client

This package was created to help JS/TS users dealing with [cep-78-enhanced-nft ](https://github.com/casper-ecosystem/cep-78-enhanced-nft). It was built on top of [casper-js-sdk](https://github.com/casper-ecosystem/casper-js-sdk)

It can be treated as a deploy builder for all of the possible interactions:

- contract installation including different configurations
- token minting
- token transfers
- token burning
- approvals
- changing configuration after installation
- setting token metada
- storing some of the contract related data in Account's `NamedKeys`

## Usage

1. To install run `npm i -S capser-cep78-js-client`.  [after it will be published]
2. import the contract in your code `import { CEP78Client } from 'capser-cep78-js-client'`
3. if you want to install it, look at the `install` method and all of the possibles configuration options (`InstallArgs`).
4. if you want to start working with already installed contract, use `setContractHash(contractHash)` method.

## Examples 

As a good starting point, you can also look into `examples/` directory to see the most common usage scenarios (`install.ts` and `usage.ts`). You can even install the contract and run the example scenario by running the `npm run example:install` and `npm run example:usage`. You will just need to provide following env variables (preferebly in a `client-js/.env`).

```
NODE_URL=http://localhost:11101/rpc
NETWORK_NAME=casper-net-1
MASTER_KEY_PAIR_PATH=/Users/someuser/.casper/casper-node/utils/nctl/assets/net-1/faucet
USER1_KEY_PAIR_PATH=/Users/someuser/.casper/casper-node/utils/nctl/assets/net-1/users/user-1
```

### CODE REVIEW

**Comment by hoffmannjan** [comment]
The last thing to fix is `migrate` method as discussed with Karan.

File: client-js/examples/usage.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  let accountInfo = await getAccountInfo(NODE_URL!, FAUCET_KEYS.publicKey);
+
+  console.log(`\n=====================================\n`);
+
+  console.log(`... Account Info: `);
+  console.log(JSON.stringify(accountInfo, null, 2));
+
+  const contractHash = await getAccountNamedKeyValue(
+    accountInfo,
+    `nft_contract`
```
Comment (gRoussac) [inline client-js/examples/usage.ts:51]: You can use double quotes if you don't need literals.

File: client-js/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import { CLValueParsers } from "casper-js-sdk";
+
+const isObject = (item: unknown) => typeof item === "object";
```
Comment (gRoussac) [inline client-js/src/utils.ts]: Seems to be a bit of overkill to have this as a util. Were do you use it ?

Comment (hoffmannjan) [reply client-js/src/utils.ts]: Right - I missed it - removed. Thx!

File: client-js/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+It can be treated as a deploy builder for all of the possible interactions:
+
+- contract installation including different configurations
+- token minting
+- token transfers
+- token burning
+- approvals
+- changing configuration after installation
+- setting token metada
```
Comment (ACStone-MTS) [inline client-js/README.md]:
```suggestion
- setting token metadata
```

File: client-js/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+## casper-cep78-js-client
+
+This package was created to help JS/TS users dealing with [cep-78-enhanced-nft ](https://github.com/casper-ecosystem/cep-78-enhanced-nft). It was built on top of [casper-js-sdk](https://github.com/casper-ecosystem/casper-js-sdk)
```
Comment (ACStone-MTS) [inline client-js/README.md]:
```suggestion
This package was created to help JS/TS users dealing with [cep-78-enhanced-nft](https://github.com/casper-ecosystem/cep-78-enhanced-nft). It was built on top of [casper-js-sdk](https://github.com/casper-ecosystem/casper-js-sdk)
```

File: client-js/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+- token transfers
+- token burning
+- approvals
+- changing configuration after installation
+- setting token metada
+- storing some of the contract related data in Account's `NamedKeys`
+
+## Usage
+
+1. To install run `npm i -S capser-cep78-js-client`.
```
Comment (ACStone-MTS) [inline client-js/README.md]:
```suggestion
1. To install, run `npm i -S capser-cep78-js-client`. 
```

File: client-js/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+- token burning
+- approvals
+- changing configuration after installation
+- setting token metada
+- storing some of the contract related data in Account's `NamedKeys`
+
+## Usage
+
+1. To install run `npm i -S capser-cep78-js-client`. 
+2. import the contract in your code `import { CEP78Client } from 'capser-cep78-js-client'`
```
Comment (ACStone-MTS) [inline client-js/README.md]:
```suggestion
2. Import the contract in your code `import { CEP78Client } from 'capser-cep78-js-client'`
```

File: client-js/README.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+3. if you want to install it, look at the `install` method and all of the possibles configuration options (`InstallArgs`).
```
Comment (ACStone-MTS) [inline client-js/README.md]:
```suggestion
3. If you want to install it, look at the `install` method and all of the possible configuration options (`InstallArgs`).
```

File: client-js/README.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+4. if you want to start working with already installed contract, use `setContractHash(contractHash)` method.
```
Comment (ACStone-MTS) [inline client-js/README.md]:
```suggestion
4. If you want to start working with a previously installed contract, use the `setContractHash(contractHash)` method.
```

File: client-js/README.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+
+## Examples 
+
+As a good starting point, you can also look into `examples/` directory to see the most common usage scenarios (`install.ts` and `usage.ts`). You can even install the contract and run the example scenario by running the `npm run example:install` and `npm run example:usage`. You will just need to provide following env variables (preferebly in a `client-js/.env`).
```
Comment (ACStone-MTS) [inline client-js/README.md]:
```suggestion
As a good starting point, you can also look into the `examples/` directory to see the most common usage scenarios (`install.ts` and `usage.ts`). You can install the contract and run the example scenario by running the `npm run example:install` and `npm run example:usage`. You will just need to provide following env variables (preferebly in a `client-js/.env`).
```

**Review by ACStone-MTS — COMMENTED** [review]
Some minor suggestions for wording.

**Comment by hoffmannjan** [comment]
Thanks @ACStoneCL!

File: client-js/examples/install.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
+import {
+  CEP78Client,
+  NFTOwnershipMode,
+  NFTKind,
+  NFTMetadataKind,
+  NFTIdentifierMode,
+  MetadataMutability,
+  OwnerReverseLookupMode,
+  MintingMode,
```
Comment (darthsiroftardis) [inline client-js/examples/install.ts:9]: I think `BurnMode` is missing here, please double check the modality parity with what is on `dev`

**Review by darthsiroftardis — APPROVED** [review]

File: client-js/examples/install.ts (hunk 10 above)
Comment (hoffmannjan) [reply client-js/examples/install.ts:9]: oh I made it optional, just checking all of the modalities

Comment (hoffmannjan) [reply client-js/examples/install.ts:9]: https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/98/files#diff-f014c08f47e43401e428e6acbb68846c9bdaac6e9520385776107e718dc7dddc

here are all modalities - in TypeScript enums without assigned values starts from `0` and increment with every entry

**Pull request merged**